### PR TITLE
Improve variable-length match plan with Loop planNode

### PR DESCRIPTION
**Pull request by Aiee**
This PR addresses 2 things:

1.  Add a new planNode: UnionAllVersionVar. 
    - This node unionizes all versions of the variable in its dependency.
2. Modify variable-length match plan, and replace for loop with the Loop planNode.
    - The updated execution plan for ```MATCH (:player{name:"Tim Duncan"})-[e: serve*0..2]-(v) RETURN v``` is as below:
single_variable_length_match

### CODE REVIEW

File: src/executor/query/UnionAllVersionVarExecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        // Check if inputData is a dataset
+        if (UNLIKELY(!inputData->isDataSet())) {
+            std::stringstream ss;
+            ss << "Invalid data types of dependencies: " << inputData->type() << ".";
+            return Status::Error(ss.str());
+        }
+        inputList.emplace_back(std::move(iter));
+    }
+
+    QueryExpressionContext ctx(ectx_);
```
Comment (yixinglu) [inline src/executor/query/UnionAllVersionVarExecutor.cpp:42]: Where to use this `ctx` ?

File: src/planner/match/Expand.cpp [hunk 2]
```diff
@@ -110,34 +112,58 @@ Status Expand::expandSteps(const NodeInfo& node,
...
+    PlanNode* firstStep = subplan.root;
+
+    // Build Start node from first step
+    SubPlan loopBodyPlan;
+    PlanNode* startNode = StartNode::make(matchCtx_->qctx);
+    startNode->setOutputVar(firstStep->outputVar());
+    startNode->setColNames(firstStep->colNames());
+    loopBodyPlan.tail = startNode;
+    loopBodyPlan.root = startNode;
+    PlanNode* passThrough = loopBodyPlan.root;
```
Comment (yixinglu) [inline src/planner/match/Expand.cpp:135]: delete this useless variable `passThrough` and use startNode directly in following function procedure.

File: src/planner/match/Expand.h [hunk 3]
```diff
@@ -52,17 +52,16 @@ class Expand final {
                       const Expression* nodeFilter,
                       SubPlan* plan);
 
-    Status collectData(const PlanNode* joinLeft,
+    Status collectData(PlanNode* joinLeft,
```
Comment (yixinglu) [inline src/planner/match/Expand.h:55]: why to delete the const qualifier?

**Review by yixinglu — COMMENTED** [review]
excellent

**Review by yixinglu — APPROVED** [review]

**Review by jievince — APPROVED** [review]
Well Done!!

**Pull request merged**